x/payment/keeper/stream_record.go

```diff
@@ -245,72 +279,260 @@ func (k Keeper) AutoSettle(ctx sdk.Context) {
 	currentTimestamp := ctx.BlockTime().Unix()
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AutoSettleRecordKeyPrefix)
 	iterator := storetypes.KVStorePrefixIterator(store, []byte{})
-
 	defer iterator.Close()
 
-	var num uint64 = 0
-	maxNum := k.GetParams(ctx).MaxAutoForceSettleNum
+	count := uint64(0)
+	max := k.GetParams(ctx).MaxAutoSettleFlowCount
 	for ; iterator.Valid(); iterator.Next() {
-		if num >= maxNum {
+		if count >= max {
 			return
 		}
-		val := types.ParseAutoSettleRecordKey(iterator.Key())
-		addr := sdk.MustAccAddressFromHex(val.Addr)
-		if val.Timestamp > currentTimestamp {
+		record := types.ParseAutoSettleRecordKey(iterator.Key())
+		addr := sdk.MustAccAddressFromHex(record.Addr)
+		if record.Timestamp > currentTimestamp {
 			return
 		}
 		streamRecord, found := k.GetStreamRecord(ctx, addr)
-		if !found {
-			ctx.Logger().Error("stream record not found", "addr", val.Addr)
-			panic("stream record not found")
+		if !found { // should not happen
+			ctx.Logger().Error("auto settle, stream record not found", "address", record.Addr)
+			continue
 		}
-		change := types.NewDefaultStreamRecordChangeWithAddr(addr)
-		err := k.UpdateStreamRecord(ctx, streamRecord, change, true)
-		if err != nil {
-			ctx.Logger().Error("force settle failed", "addr", val.Addr, "err", err)
-			panic("force settle failed")
+
+		if streamRecord.Status == types.STREAM_ACCOUNT_STATUS_ACTIVE {
+			count++ // add one for a stream record
+			err := k.SettleStreamRecord(ctx, streamRecord)
+			if err != nil {
+				ctx.Logger().Error("auto settle, settle stream record failed", "err", err.Error())
+				continue
+			}
+			k.SetStreamRecord(ctx, streamRecord)
+			if streamRecord.Status == types.STREAM_ACCOUNT_STATUS_ACTIVE {
+				continue
+			}
+			if count >= max {
+				return
+			}
+		}
+
+		if k.ExistsAutoResumeRecord(ctx, record.Timestamp, addr) { // this check should be cheap usually
```

Comment:
Is it possible that a streamrecord is `STREAM_ACCOUNT_STATUS_ACTIVE` and still `ExistsAutoResumeRecord`?

Reply:
no, when the stream is changed to `active`, the related `AutoResumeRecord` will be deleted.

Reply:
If this is the case, why dont we use the status of stream record to replace the `ExistsAutoResumeRecord ` condition here?